File: salauddin_sifat/jan26/704_binary_search.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(search([]int{-1, 0, 3, 5, 9, 12}, 9)) // Output: 4
}

func search(nums []int, target int) int {
	li := 0
	hi := len(nums) - 1
	mi := (hi + li) / 2
	for li <= hi {
		if nums[mi] == target {
			return mi
		} else if nums[mi] > target {
			hi = mi - 1
		} else {
			li = mi + 1
		}
		mi = (li + hi) / 2
	}
	if nums[mi] == target {
		return mi
	}
	return -1
}

File: salauddin_sifat/jan26/704_binary_search.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function (nums, target) {
  let li = 0;
  let hi = nums.length - 1;
  let mi = Math.floor((li + hi) / 2);
  while (li <= hi) {
    if (target === nums[mi]) {
      return mi;
    } else if (target > nums[mi]) {
      li = mi + 1;
    } else {
      hi = mi - 1;
    }
    mi = Math.floor((li + hi) / 2);
  }
  return nums[mi] === target ? mi : -1;
};
console.log(search([-1, 0, 3, 5, 9, 12], 9)); //4
